File: python/src/wh40kdc/export/yellowscribe.py
"""Yellowscribe (.ros) serializer — Python mirror of ``tools/src/export/yellowscribe.ts``.

Emits a BattleScribe-compatible ``.ros`` XML document that Yellowscribe
(github.com/ThePants999/Yellowscribe) ingests to build an army in Tabletop
Simulator.

Unlike every other exporter this one is **Dataset-backed**: the Roster carries
only entity ids/counts/points, but Yellowscribe needs full datasheet stat lines,
weapon profiles, keywords, and ability text for its TTS tooltips. So this
serializer resolves each unit against the Dataset and reads stats/weapons/
abilities off the linked views. Output is byte-identical to the TS oracle,
pinned by ``conformance/roster/*/expected.yellowscribe.ros``.

**IP boundary.** No GW rules prose is emitted — ability descriptions come from
the conformance-pinned DSL describer (:meth:`AbilityView.describe`); the dataset
stores no rules text. Everything else is a numeric fact or a community name.

See the TS module for the full Yellowscribe ``.ros`` contract and determinism
rules (no sorting, fixed attribute order, 2-space indent + LF, deterministic
``entityId + index`` ids, integer stats plain / string stats verbatim, one
shared XML escaper).
"""

from __future__ import annotations

from typing import Any

from wh40kdc.data.dataset import Dataset
from wh40kdc.export.helpers import title_case_id

#: BattleScribe's Warhammer 40,000 10th-edition game-system id — Yellowscribe
#: rejects a roster whose ``gameSystemId`` isn't this.
_GAME_SYSTEM_ID = "sys-352e-adc2-7639-d6a9"
_GAME_SYSTEM_NAME = "Warhammer 40,000"


# --------------------------------------------------------------------------
# Minimal deterministic XML tree + renderer (no library — a library would
# reorder attributes or normalise whitespace, breaking byte-parity).
# --------------------------------------------------------------------------


class _El:
    __slots__ = ("attrs", "children", "tag", "text")

    def __init__(
        self,
        tag: str,
        attrs: list[tuple[str, str]],
        children: list[_El],
        text: str | None = None,
    ) -> None:
        self.tag = tag
        self.attrs = attrs
        self.children = children
        self.text = text


def _el(tag: str, attrs: list[tuple[str, str]], children: list[_El]) -> _El:
    return _El(tag, attrs, children)


def _leaf(tag: str, attrs: list[tuple[str, str]], text: str) -> _El:
    return _El(tag, attrs, [], text)


def _esc_text(s: str) -> str:
    """Escape text content: ``& < >``."""
    return s.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")


def _esc_attr(s: str) -> str:
    """Escape an attribute value: ``& < > "``."""
    return _esc_text(s).replace('"', "&quot;")


def _render_attrs(attrs: list[tuple[str, str]]) -> str:
    return "".join(f' {k}="{_esc_attr(v)}"' for k, v in attrs)


def _render(node: _El, depth: int) -> str:
    indent = "  " * depth
    open_ = f"<{node.tag}{_render_attrs(node.attrs)}"
    if node.text is not None:
        return f"{indent}{open_}>{_esc_text(node.text)}</{node.tag}>"
    if not node.children:
        return f"{indent}{open_}/>"
    inner = "\n".join(_render(c, depth + 1) for c in node.children)
    return f"{indent}{open_}>\n{inner}\n{indent}</{node.tag}>"


# --------------------------------------------------------------------------
# Stat-line rendering (datasheet conventions; deterministic across ports).
# --------------------------------------------------------------------------


def _fmt_move(m: Any) -> str:
    """Movement: append the inch mark unless the stored value already has one."""
    s = str(m)
    return s if s.endswith('"') else f'{s}"'


def _fmt_target(v: Any) -> str:
    """A target-number stat (Sv, Ld, BS, WS): append ``+``."""
    return f"{v}+"


def _keyword_label(name: str, parameters: dict[str, Any] | None) -> str:
    """A weapon keyword's display label: ``Anti-Infantry 4+``, ``Rapid Fire 1``,
    or a bare ``Devastating Wounds``."""
    if parameters:
        tk = parameters.get("target_keyword")
        th = parameters.get("threshold")
        if isinstance(tk, str) and isinstance(th, (int, str)) and not isinstance(th, bool):
            return f"{name}-{tk} {th}+"
        value = parameters.get("value")
        if value is not None:
            return f"{name} {value}"
    return name


# --------------------------------------------------------------------------
# Profile builders.
# --------------------------------------------------------------------------


def _unit_stat_profiles(view: Any) -> list[_El]:
    """The ``<profile typeName="Unit">`` stat line(s) — one per unit stat profile
    (degrading/wound-track units carry several)."""
    out: list[_El] = []
    for i, p in enumerate(view.raw["profiles"]):
        name = p.get("name") or (view.name if i == 0 else f"{view.name} ({i + 1})")
        out.append(
            _el(
                "profile",
                [("name", name), ("typeName", "Unit")],
                [
                    _el(
                        "characteristics",
                        [],
                        [
                            _leaf("characteristic", [("name", "M")], _fmt_move(p["M"])),
                            _leaf("characteristic", [("name", "T")], str(p["T"])),
                            _leaf("characteristic", [("name", "SV")], _fmt_target(p["Sv"])),
                            _leaf("characteristic", [("name", "W")], str(p["W"])),
                            _leaf("characteristic", [("name", "LD")], _fmt_target(p["Ld"])),
                            _leaf("characteristic", [("name", "OC")], str(p["OC"])),
                        ],
                    ),
                ],
            )
        )
    return out


def _ability_profile(name: str, description: str) -> _El:
    return _el(
        "profile",
        [("name", name), ("typeName", "Abilities")],
        [
            _el(
                "characteristics",
                [],
                [
                    _leaf("characteristic", [("name", "Description")], description),
                ],
            ),
        ],
    )


def _ability_profiles(view: Any) -> list[_El]:
    """``<profile typeName="Abilities">`` entries: the invuln save (a numeric fact)
    followed by each ability's describer-rendered text."""
    out: list[_El] = []
    invuln = view.profile_at(0).get("invuln_sv")
    if invuln is not None:
        out.append(_ability_profile("Invulnerable Save", f"{invuln}+ invulnerable save"))
    for ability in view.abilities:
        out.append(_ability_profile(ability.name, ability.describe()))
    return out


def _weapon_profiles(weapon: Any) -> list[_El]:
    """A weapon's ``<profile>`` list — one per weapon stat profile. Ranged weapons
    carry ``BS``, melee carry ``WS`` and a ``Melee`` range."""
    ranged = weapon.raw["type"] == "ranged"
    type_name = "Ranged Weapons" if ranged else "Melee Weapons"
    out: list[_El] = []
    for i, p in enumerate(weapon.raw["profiles"]):
        stats = p["stats"]
        rng = _fmt_move(p.get("range") if p.get("range") is not None else 0) if ranged else "Melee"
        skill_name = "BS" if ranged else "WS"
        skill = stats.get("BS") if ranged else stats.get("WS")
        keywords = ", ".join(
            _keyword_label(k["keyword"].name, k["parameters"]) for k in weapon.keywords_at(i)
        )
        out.append(
            _el(
                "profile",
                [("name", p["name"]), ("typeName", type_name)],
                [
                    _el(
                        "characteristics",
                        [],
                        [
                            _leaf("characteristic", [("name", "Range")], rng),
                            _leaf("characteristic", [("name", "A")], str(stats["A"])),
                            _leaf(
                                "characteristic",
                                [("name", skill_name)],
                                _fmt_target(skill)
                                if isinstance(skill, int) and not isinstance(skill, bool)
                                else "N/A",
                            ),
                            _leaf("characteristic", [("name", "S")], str(stats["S"])),
                            _leaf("characteristic", [("name", "AP")], str(stats["AP"])),
                            _leaf("characteristic", [("name", "D")], str(stats["D"])),
                            _leaf("characteristic", [("name", "Keywords")], keywords),
                        ],
                    ),
                ],
            )
        )
    return out


# --------------------------------------------------------------------------
# Selection tree.
# --------------------------------------------------------------------------


def _resolve_unit(unit: dict[str, Any], dataset: Dataset, faction_id: str | None) -> Any | None:
    uid = unit["ref"]["id"]
    if uid is None:
        return None
    if faction_id:
        scoped = dataset.units.get_in_faction(uid, faction_id)
        if scoped is not None:
            return scoped
    return dataset.units.get_any(uid)


def _resolve_weapon(w: dict[str, Any], dataset: Dataset, faction_id: str | None) -> Any | None:
    wid = w["ref"]["id"]
    if wid is None:
        return None
    scoped = dataset.weapons.get_in_faction(wid, faction_id) if faction_id else None
    return scoped or dataset.weapons.get_any(wid)


def _upgrade_selection(id_: str, weapon: Any, total_count: int) -> _El:
    """One weapon ``<selection type="upgrade">``. ``number`` is the TOTAL across
    the group's models (Yellowscribe divides it back out by the model count)."""
    return _el(
        "selection",
        [("id", id_), ("name", weapon.name), ("type", "upgrade"), ("number", str(total_count))],
        [_el("profiles", [], _weapon_profiles(weapon))],
    )


def _model_selection(
    id_base: str,
    model_name: str,
    model_count: int,
    wargear: list[dict[str, Any]],
    dataset: Dataset,
    faction_id: str | None,
) -> _El:
    upgrades: list[_El] = []
    for wi, w in enumerate(wargear):
        weapon = _resolve_weapon(w, dataset, faction_id)
        if weapon is None:
            continue  # unresolved weapon — skip (already flagged in diagnostics)
        upgrades.append(_upgrade_selection(f"{id_base}-w{wi}", weapon, w["count"] * model_count))
    children: list[_El] = []
    if upgrades:
        children.append(_el("selections", [], upgrades))
    return _el(
        "selection",
        [("id", id_base), ("name", model_name), ("type", "model"), ("number", str(model_count))],
        children,
    )


def _model_selections(
    unit: dict[str, Any], unit_id: str, view: Any, dataset: Dataset, faction_id: str | None
) -> list[_El]:
    """One ``<selection type="model">`` per loadout group, falling back to a single
    group over the flat ``wargear`` (whose counts are already unit totals)."""
    groups = unit.get("loadout_groups")
    if not groups:
        groups = [{"model_name": None, "count": unit["model_count"], "wargear": unit["wargear"]}]
    return [
        _model_selection(
            f"{unit_id}-m{gi}",
            g["model_name"] or view.name,
            g["count"],
            g["wargear"],
            dataset,
            faction_id,
        )
        for gi, g in enumerate(groups)
    ]


def _categories_el(view: Any) -> _El | None:
    """Unit categories: faction keywords (prefixed ``Faction: ``) then general
    keywords, in stored order."""
    cats: list[_El] = []
    for k in view.raw.get("faction_keywords") or []:
        cats.append(_el("category", [("name", f"Faction: {k}")], []))
    for k in view.raw.get("keywords") or []:
        cats.append(_el("category", [("name", k)], []))
    return _el("categories", [], cats) if cats else None


def _unit_selection(
    unit: dict[str, Any], index: int, dataset: Dataset, faction_id: str | None
) -> _El | None:
    """One unit ``<selection type="unit">``. ``None`` for a unit that doesn't
    resolve (no datasheet to emit stats from)."""
    view = _resolve_unit(unit, dataset, faction_id)
    if view is None:
        return None
    unit_id = f"unit{index}"

    profiles = _unit_stat_profiles(view) + _ability_profiles(view)
    children: list[_El] = [_el("profiles", [], profiles)]

    cats = _categories_el(view)
    if cats is not None:
        children.append(cats)

    children.append(
        _el("selections", [], _model_selections(unit, unit_id, view, dataset, faction_id))
    )

    return _el(
        "selection",
        [("id", unit_id), ("name", unit["ref"]["raw_name"]), ("type", "unit"), ("number", "1")],
        children,
    )


def serialize_yellowscribe(roster: dict[str, Any], dataset: Dataset) -> str:
    """Serialize a Roster into Yellowscribe-ingestible BattleScribe ``.ros`` XML."""
    faction_id = roster.get("faction_id")
    faction_name = title_case_id(faction_id) or "Unknown"

    unit_selections: list[_El] = []
    for i, unit in enumerate(roster["units"]):
        sel = _unit_selection(unit, i, dataset, faction_id)
        if sel is not None:
            unit_selections.append(sel)

    force = _el(
        "force",
        [("id", "force0"), ("name", faction_name), ("catalogueName", faction_name)],
        [_el("selections", [], unit_selections)],
    )

    roster_el = _el(
        "roster",
        [
            ("id", "roster0"),
            ("name", roster["name"]),
            ("gameSystemId", _GAME_SYSTEM_ID),
            ("gameSystemName", _GAME_SYSTEM_NAME),
        ],
        [_el("forces", [], [force])],
    )

    return f'<?xml version="1.0" encoding="utf-8"?>\n{_render(roster_el, 0)}\n'

File: crates/wh40kdc/src/export/yellowscribe.rs
//! Yellowscribe serializer (`yellowscribe`) — emits a BattleScribe-compatible
//! `.ros` XML document that Yellowscribe (github.com/ThePants999/Yellowscribe)
//! ingests to build an army in Tabletop Simulator.
//!
//! Unlike every other Rust exporter this one is **Dataset-backed**: the
//! [`Roster`] carries only entity ids, counts, and points, but Yellowscribe
//! needs full datasheet stat lines, weapon profiles, keywords, and ability text
//! for its TTS tooltips and mini-datasheets. So this serializer resolves each
//! unit against the [`Dataset`] (faction-first, then any) and reads the stats /
//! weapons / abilities off the linked records. It is reached through
//! [`export_roster_with_dataset`](super::export_roster_with_dataset) rather than
//! the Dataset-free [`export_roster`](super::export_roster), so the other
//! formats are untouched.
//!
//! **IP boundary.** No GW rules prose is ever emitted. Ability descriptions come
//! from the conformance-pinned DSL describer
//! ([`describe_ability`](crate::translate::describe_ability)); the dataset stores
//! no rules text. Everything else is a numeric fact (stat lines) or a
//! community-authored name.
//!
//! **Determinism** (byte-identical across the TS/Rust/Python/Go ports for
//! conformance): no sorting — units in `roster.units` order, models in
//! `loadout_groups` order, weapons/keywords/abilities in their stored array
//! order; fixed attribute order; fixed 2-space indent + LF; deterministic
//! synthetic ids (`unit{index}`, `unit{i}-m{g}`, `unit{i}-m{g}-w{w}`); integer
//! stats plain, string stats verbatim; one shared XML escaper.
//!
//! Rust mirror of `tools/src/export/yellowscribe.ts` (and the Python
//! `python/src/wh40kdc/export/yellowscribe.py`).

use crate::data::Dataset;
use crate::generated::{
    Unit, Weapon, WeaponProfilesItem, WeaponProfilesItemKeywordsItemParameters, WeaponType,
};
use crate::import::{Roster, RosterLoadoutGroup, RosterUnit, RosterWargear};
use crate::translate::describe_ability;

use super::helpers::title_case_id;

/// BattleScribe's Warhammer 40,000 10th-edition game-system id — Yellowscribe
/// rejects a roster whose `gameSystemId` isn't this.
const GAME_SYSTEM_ID: &str = "sys-352e-adc2-7639-d6a9";
const GAME_SYSTEM_NAME: &str = "Warhammer 40,000";

// ---------------------------------------------------------------------------
// Minimal deterministic XML tree + renderer (no library — a library would
// reorder attributes or normalise whitespace, breaking byte-parity).
// ---------------------------------------------------------------------------

/// A single XML element. Attributes are held in fixed emission order (never a
/// map/hash). `text` and `children` are mutually exclusive: a `text` node is a
/// leaf whose body is escaped character data; a `children` node nests further
/// elements; an empty node (no text, no children) renders self-closing.
struct XmlEl {
    tag: &'static str,
    attrs: Vec<(&'static str, String)>,
    children: Vec<XmlEl>,
    text: Option<String>,
}

fn el(tag: &'static str, attrs: Vec<(&'static str, String)>, children: Vec<XmlEl>) -> XmlEl {
    XmlEl {
        tag,
        attrs,
        children,
        text: None,
    }
}

fn leaf(tag: &'static str, attrs: Vec<(&'static str, String)>, text: String) -> XmlEl {
    XmlEl {
        tag,
        attrs,
        children: Vec::new(),
        text: Some(text),
    }
}

/// Escape text content: `& < >`.
fn esc_text(s: &str) -> String {
    s.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
}

/// Escape an attribute value: `& < > "`.
fn esc_attr(s: &str) -> String {
    esc_text(s).replace('"', "&quot;")
}

fn render_attrs(attrs: &[(&'static str, String)]) -> String {
    attrs
        .iter()
        .map(|(k, v)| format!(" {k}=\"{}\"", esc_attr(v)))
        .collect()
}

fn render(node: &XmlEl, depth: usize) -> String {
    let indent = "  ".repeat(depth);
    let open = format!("<{}{}", node.tag, render_attrs(&node.attrs));
    if let Some(text) = &node.text {
        return format!("{indent}{open}>{}</{}>", esc_text(text), node.tag);
    }
    if node.children.is_empty() {
        return format!("{indent}{open}/>");
    }
    let inner = node
        .children
        .iter()
        .map(|c| render(c, depth + 1))
        .collect::<Vec<_>>()
        .join("\n");
    format!("{indent}{open}>\n{inner}\n{indent}</{}>", node.tag)
}

// ---------------------------------------------------------------------------
// Stat-line rendering (datasheet conventions; deterministic across ports).
// ---------------------------------------------------------------------------

/// Movement: append the inch mark unless the stored value already carries one.
fn fmt_move(s: &str) -> String {
    if s.ends_with('"') {
        s.to_string()
    } else {
        format!("{s}\"")
    }
}

/// A "target-number" stat (Sv, Ld, BS, WS): append `+`.
fn fmt_target(v: i64) -> String {
    format!("{v}+")
}

/// A weapon keyword's display label: `Anti-Infantry 4+`, `Rapid Fire 1`, or a
/// bare `Devastating Wounds`. Mirrors the datasheet convention.
fn keyword_label(
    name: &str,
    parameters: Option<&WeaponProfilesItemKeywordsItemParameters>,
) -> String {
    if let Some(p) = parameters {
        if let (Some(tk), Some(th)) = (p.target_keyword.as_ref(), p.threshold) {
            return format!("{name}-{} {th}+", tk.as_str());
        }
        if let Some(value) = p.value.as_ref() {
            return format!("{name} {value}");
        }
    }
    name.to_string()
}

// ---------------------------------------------------------------------------
// Profile builders.
// ---------------------------------------------------------------------------

/// The `<profile typeName="Unit">` stat line(s). Emits one profile per unit stat
/// profile (degrading/wound-track units carry several).
fn unit_stat_profiles(unit: &Unit) -> Vec<XmlEl> {
    unit.profiles
        .iter()
        .enumerate()
        .map(|(i, p)| {
            let name = p.name.clone().unwrap_or_else(|| {
                if i == 0 {
                    unit.name.as_str().to_string()
                } else {
                    format!("{} ({})", unit.name.as_str(), i + 1)
                }
            });
            el(
                "profile",
                vec![("name", name), ("typeName", "Unit".to_string())],
                vec![el(
                    "characteristics",
                    Vec::new(),
                    vec![
                        leaf(
                            "characteristic",
                            vec![("name", "M".to_string())],
                            fmt_move(&p.m.to_string()),
                        ),
                        leaf(
                            "characteristic",
                            vec![("name", "T".to_string())],
                            p.t.to_string(),
                        ),
                        leaf(
                            "characteristic",
                            vec![("name", "SV".to_string())],
                            fmt_target(p.sv),
                        ),
                        leaf(
                            "characteristic",
                            vec![("name", "W".to_string())],
                            p.w.to_string(),
                        ),
                        leaf(
                            "characteristic",
                            vec![("name", "LD".to_string())],
                            fmt_target(p.ld),
                        ),
                        leaf(
                            "characteristic",
                            vec![("name", "OC".to_string())],
                            p.oc.to_string(),
                        ),
                    ],
                )],
            )
        })
        .collect()
}

fn ability_profile(name: &str, description: &str) -> XmlEl {
    el(
        "profile",
        vec![
            ("name", name.to_string()),
            ("typeName", "Abilities".to_string()),
        ],
        vec![el(
            "characteristics",
            Vec::new(),
            vec![leaf(
                "characteristic",
                vec![("name", "Description".to_string())],
                description.to_string(),
            )],
        )],
    )
}

/// `<profile typeName="Abilities">` entries: the invuln save (a numeric fact)
/// followed by each ability's describer-rendered text.
fn ability_profiles(unit: &Unit, dataset: &Dataset) -> Vec<XmlEl> {
    let mut out: Vec<XmlEl> = Vec::new();
    if let Some(invuln) = unit.profiles.first().and_then(|p| p.invuln_sv) {
        out.push(ability_profile(
            "Invulnerable Save",
            &format!("{invuln}+ invulnerable save"),
        ));
    }
    for ability in dataset.abilities_of(unit) {
        out.push(ability_profile(&ability.name, &describe_ability(ability)));
    }
    out
}

/// A weapon's `<profile>` list — one per weapon stat profile (e.g. a plasma
/// gun's standard / supercharge). Ranged weapons carry `BS`, melee carry `WS`
/// and a `Melee` range.
fn weapon_profiles(weapon: &Weapon, dataset: &Dataset) -> Vec<XmlEl> {
    let ranged = weapon.type_ == WeaponType::Ranged;
    let type_name = if ranged {
        "Ranged Weapons"
    } else {
        "Melee Weapons"
    };
    weapon
        .profiles
        .iter()
        .map(|p| {
            let stats = &p.stats;
            let range = if ranged {
                let raw = p
                    .range
                    .as_ref()
                    .map(|r| r.to_string())
                    .unwrap_or_else(|| "0".to_string());
                fmt_move(&raw)
            } else {
                "Melee".to_string()
            };
            let skill_name = if ranged { "BS" } else { "WS" };
            let skill = if ranged { stats.bs } else { stats.ws };
            let skill_text = match skill {
                Some(v) => fmt_target(v),
                None => "N/A".to_string(),
            };
            let keywords = keywords_label(p, dataset);
            el(
                "profile",
                vec![
                    ("name", p.name.as_str().to_string()),
                    ("typeName", type_name.to_string()),
                ],
                vec![el(
                    "characteristics",
                    Vec::new(),
                    vec![
                        leaf("characteristic", vec![("name", "Range".to_string())], range),
                        leaf(
                            "characteristic",
                            vec![("name", "A".to_string())],
                            stats.a.to_string(),
                        ),
                        leaf(
                            "characteristic",
                            vec![("name", skill_name.to_string())],
                            skill_text,
                        ),
                        leaf(
                            "characteristic",
                            vec![("name", "S".to_string())],
                            stats.s.to_string(),
                        ),
                        leaf(
                            "characteristic",
                            vec![("name", "AP".to_string())],
                            stats.ap.to_string(),
                        ),
                        leaf(
                            "characteristic",
                            vec![("name", "D".to_string())],
                            stats.d.to_string(),
                        ),
                        leaf(
                            "characteristic",
                            vec![("name", "Keywords".to_string())],
                            keywords,
                        ),
                    ],
                )],
            )
        })
        .collect()
}

/// The comma-joined keyword labels for a single weapon profile. Unresolved
/// keyword ids are skipped (mirrors `WeaponView.keywordsAt`).
fn keywords_label(profile: &WeaponProfilesItem, dataset: &Dataset) -> String {
    profile
        .keywords
        .iter()
        .filter_map(|ref_| {
            dataset
                .weapon_keywords
                .get(ref_.keyword_id.as_str())
                .map(|kw| keyword_label(kw.name.as_str(), ref_.parameters.as_ref()))
        })
        .collect::<Vec<_>>()
        .join(", ")
}

// ---------------------------------------------------------------------------
// Selection tree.
// ---------------------------------------------------------------------------

/// Resolve a roster unit to its datasheet, faction-first (matching how a
/// `UnitView` resolves its own ids).
fn resolve_unit<'d>(
    unit: &RosterUnit,
    dataset: &'d Dataset,
    faction: Option<&str>,
) -> Option<&'d Unit> {
    let id = unit.ref_.id.as_deref()?;
    if let Some(f) = faction {
        if let Some(u) = dataset.units.get_in_faction(id, f) {
            return Some(u);
        }
    }
    dataset.units.get_any(id)
}

/// Resolve a wargear ref to its weapon, faction-first (matching how a
/// `UnitView` resolves its own weapon ids).
fn resolve_weapon<'d>(
    w: &RosterWargear,
    dataset: &'d Dataset,
    faction: Option<&str>,
) -> Option<&'d Weapon> {
    let id = w.ref_.id.as_deref()?;
    if let Some(f) = faction {
        if let Some(weapon) = dataset.weapons.get_in_faction(id, f) {
            return Some(weapon);
        }
    }
    dataset.weapons.get_any(id)
}

/// One weapon `<selection type="upgrade">`. `number` is the TOTAL across the
/// group's models (`perModel × groupModelCount`) — Yellowscribe divides it back
/// out by the model count.
fn upgrade_selection(id: String, weapon: &Weapon, total_count: u64, dataset: &Dataset) -> XmlEl {
    el(
        "selection",
        vec![
            ("id", id),
            ("name", weapon.name.as_str().to_string()),
            ("type", "upgrade".to_string()),
            ("number", total_count.to_string()),
        ],
        vec![el("profiles", Vec::new(), weapon_profiles(weapon, dataset))],
    )
}

/// One `<selection type="model">` for a loadout group, with its per-model
/// weapons nested as upgrade selections.
fn model_selection(
    id_base: &str,
    model_name: &str,
    model_count: u64,
    wargear: &[RosterWargear],
    dataset: &Dataset,
    faction: Option<&str>,
) -> XmlEl {
    let mut upgrades: Vec<XmlEl> = Vec::new();
    for (wi, w) in wargear.iter().enumerate() {
        let Some(weapon) = resolve_weapon(w, dataset, faction) else {
            continue; // unresolved weapon — skip (already flagged in diagnostics)
        };
        upgrades.push(upgrade_selection(
            format!("{id_base}-w{wi}"),
            weapon,
            w.count * model_count,
            dataset,
        ));
    }
    let mut children: Vec<XmlEl> = Vec::new();
    if !upgrades.is_empty() {
        children.push(el("selections", Vec::new(), upgrades));
    }
    el(
        "selection",
        vec![
            ("id", id_base.to_string()),
            ("name", model_name.to_string()),
            ("type", "model".to_string()),
            ("number", model_count.to_string()),
        ],
        children,
    )
}

/// The nested `<selection type="model">` list for a unit — one per loadout
/// group, falling back to a single group over the flat `wargear` (whose counts
/// are already unit totals, so per-model = total / model_count, as Yellowscribe
/// expects).
fn model_selections(
    unit: &RosterUnit,
    unit_id: &str,
    view: &Unit,
    dataset: &Dataset,
    faction: Option<&str>,
) -> Vec<XmlEl> {
    // Borrow the real groups when present, else synthesize a single fallback
    // group over the flat wargear.
    let fallback: [RosterLoadoutGroup; 1] = [RosterLoadoutGroup {
        model_name: None,
        count: unit.model_count,
        wargear: unit.wargear.clone(),
    }];
    let groups: &[RosterLoadoutGroup] = match unit.loadout_groups.as_ref() {
        Some(g) if !g.is_empty() => g.as_slice(),
        _ => &fallback,
    };
    groups
        .iter()
        .enumerate()
        .map(|(gi, g)| {
            let model_name = g
                .model_name
                .as_deref()
                .unwrap_or_else(|| view.name.as_str());
            model_selection(
                &format!("{unit_id}-m{gi}"),
                model_name,
                g.count,
                &g.wargear,
                dataset,
                faction,
            )
        })
        .collect()
}

/// The unit categories (`<category>`): faction keywords (prefixed `"Faction: "`)
/// then general keywords, in stored order.
fn categories_el(view: &Unit) -> Option<XmlEl> {
    let mut cats: Vec<XmlEl> = Vec::new();
    if let Some(fk) = view.faction_keywords.as_ref() {
        for k in fk.iter() {
            cats.push(el(
                "category",
                vec![("name", format!("Faction: {}", k.as_str()))],
                Vec::new(),
            ));
        }
    }
    if let Some(kw) = view.keywords.as_ref() {
        for k in kw.iter() {
            cats.push(el(
                "category",
                vec![("name", k.as_str().to_string())],
                Vec::new(),
            ));
        }
    }
    if cats.is_empty() {
        None
    } else {
        Some(el("categories", Vec::new(), cats))
    }
}

/// One unit `<selection type="unit">`. Returns `None` for a unit that doesn't
/// resolve against the dataset (no datasheet to emit stats from).
fn unit_selection(
    unit: &RosterUnit,
    index: usize,
    dataset: &Dataset,
    faction: Option<&str>,
) -> Option<XmlEl> {
    let view = resolve_unit(unit, dataset, faction)?;
    let unit_id = format!("unit{index}");

    let mut profiles = unit_stat_profiles(view);
    profiles.extend(ability_profiles(view, dataset));
    let mut children: Vec<XmlEl> = vec![el("profiles", Vec::new(), profiles)];

    if let Some(cats) = categories_el(view) {
        children.push(cats);
    }

    let models = model_selections(unit, &unit_id, view, dataset, faction);
    children.push(el("selections", Vec::new(), models));

    Some(el(
        "selection",
        vec![
            ("id", unit_id),
            ("name", unit.ref_.raw_name.clone()),
            ("type", "unit".to_string()),
            ("number", "1".to_string()),
        ],
        children,
    ))
}

/// Serialize a [`Roster`] into Yellowscribe-ingestible BattleScribe `.ros` XML.
pub fn serialize(roster: &Roster, dataset: &Dataset) -> String {
    let faction = roster.faction_id.as_deref();
    let faction_name = title_case_id(faction).unwrap_or_else(|| "Unknown".to_string());

    let mut unit_selections: Vec<XmlEl> = Vec::new();
    for (i, unit) in roster.units.iter().enumerate() {
        if let Some(sel) = unit_selection(unit, i, dataset, faction) {
            unit_selections.push(sel);
        }
    }

    let force = el(
        "force",
        vec![
            ("id", "force0".to_string()),
            ("name", faction_name.clone()),
            ("catalogueName", faction_name),
        ],
        vec![el("selections", Vec::new(), unit_selections)],
    );

    let roster_el = el(
        "roster",
        vec![
            ("id", "roster0".to_string()),
            ("name", roster.name.clone()),
            ("gameSystemId", GAME_SYSTEM_ID.to_string()),
            ("gameSystemName", GAME_SYSTEM_NAME.to_string()),
        ],
        vec![el("forces", Vec::new(), vec![force])],
    );

    format!(
        "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n{}\n",
        render(&roster_el, 0)
    )
}
